kratos/tests/cpp_tests/utilities/test_parallel_utilities.cpp

```diff
@@ -0,0 +1,96 @@
+//    |  /           |
+//    ' /   __| _` | __|  _ \   __|
+//    . \  |   (   | |   (   |\__ `
+//   _|\_\_|  \__,_|\__|\___/ ____/
+//                   Multi-Physics
+//
+//  License:		 BSD License
+//					 Kratos default license: kratos/license.txt
+//
+//  Main authors:    Riccardo Rossi
+
+#include <utility>
+#include <iostream>
+#include "testing/testing.h"
+#include "utilities/parallel_utilities.h"
+
+namespace Kratos {
+namespace Testing {
+
+// Basic Type
+KRATOS_TEST_CASE_IN_SUITE(BlockPartitioner, KratosCoreFastSuite)
+{
+    int nsize = 1e3;
+    std::vector<double> data_vector(nsize);
+    for(auto& it : data_vector)
+        it = 5.0;
+
+    //here we raise every entry of a vector to the power 0.1
+    //c++11 version
+    BlockPartition<std::vector<double>>(data_vector).for_each(
+                                         [](double& item)
+    {
+        item = std::pow(item, 0.1);
+    });
+
+    //c++17 version - no explicit template parameter in constructor and auto in lambda
+    // BlockPartition(data_vector).for_each(
+    //     [](auto& item){
+    //         item = std::pow(item, 0.1);
+    //     });
+
+    //error check
+    for(auto& item : data_vector)
+    {
+        KRATOS_CHECK_EQUAL(item, std::pow(5.0, 0.1));
+    }
+
+    //here we check for a reduction (computing the sum of all the entries)
+    SumReduction<double> reducer;
+
+    //compute a sum reduction - c++11 version
+    BlockPartition<std::vector<double>>(data_vector).for_each(
+        reducer,
+        [](double& item, SumReduction<double>& reduction_helper)
+        {
+            reduction_helper.mvalue += item;
+        }
+    );
+
+    //compute a sum reduction - c++17 version
+    // BlockPartition(data_vector).for_each(
+    //     reducer,
+    //     [](auto& item, auto& reduction_helper){
+    //         reduction_helper.mvalue += item;
+    //         }
+    //     );
+
+    //get the value of the final sum
+    double final_sum = reducer.mvalue;
```

Comment:
I still feel this is too verbose for a reduction. Can you give an example of where this pattern would be necessary?

Reply:
ok i am now returning the value, so it should be cleaner